:book: New Provider documentation (#1682)
kubernetes-sigs/cluster-api: docs/book/src/providers/implementers-guide/building_running_and_testing.md

# Building, Running, Testing

## Docker Image Name

The patch in `config/default/manager_image_patch.yaml` will be applied to the manager pod. 
Right now there is a placeholder `IMAGE_URL`, which you will need to change to your actual image.

### Development Images
It's likely that you will want one location and tag for release development, and another during development.

> The approach most Cluster API projects is using [a `Makefile` that uses `sed` to replace the image URL][sed] on demand during development.

Reviewer: This line doesn't render correctly (see https://deploy-preview-1682--kubernetes-sigs-cluster-api.netlify.com/providers/implementers-guide/building_running_and_testing.html)

[sed]: https://github.com/kubernetes-sigs/cluster-api/blob/e0fb83a839b2755b14fbefbe6f93db9a58c76952/Makefile#L201-L204

## Deployment

> ### CertManager

Reviewer: ```suggestion
### cert-manager
```

> Cluster API uses [certmanager] to manage the certificates it needs for its webhooks. 

Reviewer: ```suggestion
Cluster API uses [cert-manager] to manage the certificates it needs for its webhooks. 
```

  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True 
```

[install]: https://cluster-api.sigs.k8s.io/tasks/installation.html#install-cluster-api

### Your provider

Now you can apply your provider as well:

```
$ cd cluster-api-provider-mailgun
$ kustomize build config/default | envsubst | kubectl apply -f-
$ kubectl describe -n cluster-api-provider-mailgun-system pod | grep -A 5 Conditions 
Conditions:
  Type              Status
  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True 
```

### Tiltfile
Cluster API development requires a lot of iteration, and the "build, tag, push, update deployment" workflow can be very tedious. 
[Tilt](https://tilt.dev) makes this process much simpler by watching for updates, then automatically building and deploying them. 

You can visit [some example repositories][capidev], but you can get started by writing out a yaml manifest and using the following [`Tiltfile`][tiltfile]
`kustomize build config/default | envsubst > capm.yaml`

[capidev]: https://github.com/chuckha/capi-dev
[tiltfile]: https://docs.tilt.dev/tiltfile_concepts.html

```starlark

> allow_k8s_contexts('kubernetes-admin@luna')

Reviewer: luna?

k8s_yaml('capm.yaml')

> docker_build('<your docke username or repository url>/cluster-api-controller-mailgun-amd64', '.')

Reviewer: ```suggestion
docker_build('<your docker username or repository url>/cluster-api-controller-mailgun-amd64', '.')
```

## Docker Image Name

The patch in `config/default/manager_image_patch.yaml` will be applied to the manager pod. 
Right now there is a placeholder `IMAGE_URL`, which you will need to change to your actual image.

### Development Images
It's likely that you will want one location and tag for release development, and another during development.

The approach most Cluster API projects is using [a `Makefile` that uses `sed` to replace the image URL][sed] on demand during development.

[sed]: https://github.com/kubernetes-sigs/cluster-api/blob/e0fb83a839b2755b14fbefbe6f93db9a58c76952/Makefile#L201-L204

## Deployment

### cert-manager

Cluster API uses [cert-manager] to manage the certificates it needs for its webhooks. 
Before you apply Cluster API's yaml, you should [install `cert-manager`][cm-install]

[cert-manager]: https://github.com/jetstack/cert-manager
[cm-install]: https://docs.cert-manager.io/en/latest/getting-started/install/kubernetes.html

```
kubectl apply -f https://github.com/jetstack/cert-manager/releases/download/<version>/cert-manager.yaml
```

### Cluster API

Before you can deploy the infrastructure controller, you'll need to deploy Cluster API itself.

> You can [use a precompiled manifest][install], or clone [`cluster-api`][capi] and apply its manifests using `kustomize`.

Reviewer 2: capi link is broken
